cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerModificationService.java

```diff
@@ -170,47 +174,125 @@ public void cleanupCluster(Telemetry telemetry) throws CloudbreakException {
             refreshParcelRepos(clouderaManagerResourceApi);
-            downloadParcel(stackProductVersion, parcelResourceApi, product);
-            distributeParcel(stackProductVersion, parcelResourceApi, product);
             if (patchUpgrade) {
-                activateParcel(stackProductVersion, parcelResourceApi, product);
+                downloadAndActivateParcels(products, parcelResourceApi);
                 restartServices(clustersResourceApi);
             } else {
-                callUpgradeCdhCommand(stackProductVersion, clustersResourceApi);
-                restartStaleServices(mgmtServiceResourceApi, clustersResourceApi);
+                ClouderaManagerProduct cdhProduct = getCdhProducts(products);
+                upgradeNonCdhProducts(products, cdhProduct.getName(), parcelResourceApi);
+                upgradeCdh(clustersResourceApi, parcelResourceApi, cdhProduct);
+                restartStaleServices(clustersResourceApi);
             }
-            clouderaManagerParcelService.removeUnusedParcelVersions(apiClient, parcelsResourceApi, parcelResourceApi, stack, product, stackProductVersion);
+            removeUnusedParcelVersions(parcelResourceApi, products);
             configService.enableKnoxAutorestartIfCmVersionAtLeast(CLOUDERAMANAGER_VERSION_7_1_0, apiClient, stack.getName());
-        } catch (ApiException | IOException e) {
+        } catch (ApiException e) {
             LOGGER.info("Could not upgrade Cloudera Runtime services", e);
             throw new ClouderaManagerOperationFailedException(e.getMessage(), e);
         }
     }
 
-    private ClusterComponent getStackComponent(Set<ClusterComponent> components) {
+    private void upgradeCdh(ClustersResourceApi clustersResourceApi, ParcelResourceApi parcelResourceApi, ClouderaManagerProduct cdhService)
+            throws ApiException, CloudbreakException {
+        downloadParcels(Collections.singleton(cdhService), parcelResourceApi);
+        distributeParcels(Collections.singleton(cdhService), parcelResourceApi);
+        callUpgradeCdhCommand(cdhService, clustersResourceApi);
+    }
+
+    private Set<ClouderaManagerProduct> getProducts(Set<ClusterComponent> components) {
         return components.stream()
-                .filter(clusterComponent -> clusterComponent.getName().equals(com.sequenceiq.cloudbreak.cloud.model.component.StackType.CDH.name()))
+                .filter(clusterComponent -> ComponentType.CDH_PRODUCT_DETAILS.equals(clusterComponent.getComponentType()))
+                .map(this::getClouderaManagerProduct)
+                .collect(Collectors.toSet());
+    }
+
+    private void upgradeNonCdhProducts(Set<ClouderaManagerProduct> services, String cdhServiceName, ParcelResourceApi parcelResourceApi)
+            throws CloudbreakException, ApiException {
+        Set<ClouderaManagerProduct> nonCdhServices = getNonCdhProducts(services, cdhServiceName);
+        if (!nonCdhServices.isEmpty()) {
```

Comment:
Can you extend these with log messages? An else branch can be introduced to log that we're skipping non-cdh based activaten. On the other case we can log what are the parcel names that we are going to activate.